File: sdk/js/test_compute_market.js
import assert from 'node:assert/strict';

import { ComputeMarketClient, Keypair } from './dist/index.js';

function encoded(values, trailing = false) {
    const bytes = new Uint8Array(values.length * 8 + (trailing ? 1 : 0));
    const view = new DataView(bytes.buffer);
    values.forEach((value, index) => view.setBigUint64(index * 8, BigInt(value), true));
    return { success: true, returnCode: 0, returnData: Buffer.from(bytes).toString('base64') };
}

const calls = [];
const fakeConnection = {
    async getSymbolRegistry(symbol) {
        if (symbol === 'COMPUTE') return { program: '11111111111111111111111111111112' };
        throw new Error('missing symbol');
    },
    async callContract(_signer, _program, functionName, args, value) {
        calls.push({ functionName, args, value });
        return 'test-signature';
    },
    async callReadonlyContract(_program, functionName) {
        if (functionName === 'get_accounting_migration_status') return encoded([2, 1, 100, 20, 0, 1]);
        if (functionName === 'get_accounting_health') return encoded([3, 0, 100, 20, 10, 130, 150, 1]);
        throw new Error(`unexpected ${functionName}`);
    },
};

const client = new ComputeMarketClient(fakeConnection);
const signer = Keypair.fromSeed(Uint8Array.from({ length: 32 }, (_, index) => index));
const migration = await client.getAccountingMigrationStatus();
const health = await client.getAccountingHealth();
assert.equal(migration.cursor, 1n);
assert.equal(migration.locked, true);
assert.equal(health.totalLiability, 130n);
assert.equal(health.solvent, true);

await client.beginAccountingV3Migration(signer, 2n);
await client.migrateAccountingV3Job(signer, 1n);
await client.completeAccountingV3Migration(signer, 100n, 20n, 10n, 130n);
assert.deepEqual(calls.map(call => call.functionName), [
    'begin_accounting_v3_migration',
    'migrate_accounting_v3_job',
    'complete_accounting_v3_migration',
]);
assert.deepEqual([...calls[0].args.slice(0, 3)], [0xAB, 0x20, 0x08]);
assert.deepEqual([...calls[1].args.slice(0, 2)], [0xAB, 0x08]);
assert.deepEqual([...calls[2].args.slice(0, 6)], [0xAB, 0x20, 0x08, 0x08, 0x08, 0x08]);

await assert.rejects(client.submitJob(signer, {
    computeUnits: 10n,
    maxPrice: 100n,
    codeHash: new Uint8Array(32),
}), /zero hash/);

const trailingClient = new ComputeMarketClient({
    ...fakeConnection,
    async callReadonlyContract(_program, functionName) {
        if (functionName === 'get_accounting_migration_status') return encoded([2, 1, 100, 20, 0, 1], true);
        if (functionName === 'get_accounting_health') return encoded([3, 0, 100, 20, 10, 130, 150, 1], true);
        throw new Error(`unexpected ${functionName}`);
    },
});
await assert.rejects(trailingClient.getAccountingMigrationStatus(), /exactly 48 bytes/);
await assert.rejects(trailingClient.getAccountingHealth(), /exactly 64 bytes/);

console.log('Compute Market SDK exact-accounting tests passed');

File: sdk/python/test_compute_market.py
from __future__ import annotations

import base64

import pytest

from lichen import ComputeMarketClient, Keypair, PublicKey


class FakeConnection:
    def __init__(self, *, append_trailing_byte: bool = False) -> None:
        self.calls: list[tuple[str, str, str, bytes, int]] = []
        self.append_trailing_byte = append_trailing_byte

    async def get_symbol_registry(self, symbol: str):
        if symbol == "COMPUTE":
            return {"program": "11111111111111111111111111111112"}
        raise RuntimeError("missing symbol")

    async def call_contract(
        self,
        caller: Keypair,
        contract: PublicKey,
        function_name: str,
        args: bytes = b"",
        value: int = 0,
    ) -> str:
        self.calls.append((str(caller.pubkey()), str(contract), function_name, args, value))
        return "test-signature"

    async def call_readonly_contract(
        self,
        contract: PublicKey,
        function_name: str,
        args: bytes = b"",
        from_pubkey: PublicKey | None = None,
    ):
        if function_name == "get_accounting_migration_status":
            values = (2, 1, 100, 20, 0, 1)
        elif function_name == "get_accounting_health":
            values = (3, 0, 100, 20, 10, 130, 150, 1)
        else:
            raise RuntimeError(f"unexpected readonly function: {function_name}")
        payload = b"".join(value.to_bytes(8, "little") for value in values)
        if self.append_trailing_byte:
            payload += b"\x00"
        return {
            "success": True,
            "returnCode": 0,
            "returnData": base64.b64encode(payload).decode("ascii"),
        }


@pytest.mark.asyncio
async def test_compute_market_accounting_v3_reads_and_writes() -> None:
    connection = FakeConnection()
    client = ComputeMarketClient(connection)
    admin = Keypair.from_seed(bytes(range(32)))

    migration = await client.get_accounting_migration_status()
    health = await client.get_accounting_health()
    await client.begin_accounting_v3_migration(admin, 2)
    await client.migrate_accounting_v3_job(admin, 1)
    await client.complete_accounting_v3_migration(admin, 100, 20, 10, 130)

    assert migration == {
        "expected_job_count": 2,
        "cursor": 1,
        "reconstructed_escrow": 100,
        "reconstructed_unpaid": 20,
        "accounting_version": 0,
        "locked": True,
    }
    assert health == {
        "accounting_version": 3,
        "migration_locked": False,
        "escrow_liability": 100,
        "unpaid_liability": 20,
        "platform_fees": 10,
        "total_liability": 130,
        "custody_balance": 150,
        "solvent": True,
    }
    assert [call[2] for call in connection.calls] == [
        "begin_accounting_v3_migration",
        "migrate_accounting_v3_job",
        "complete_accounting_v3_migration",
    ]
    assert connection.calls[0][3][:3] == bytes([0xAB, 0x20, 0x08])
    assert connection.calls[1][3][:2] == bytes([0xAB, 0x08])
    assert connection.calls[2][3][:6] == bytes([0xAB, 0x20, 0x08, 0x08, 0x08, 0x08])


@pytest.mark.asyncio
async def test_compute_market_rejects_noncanonical_read_payloads() -> None:
    client = ComputeMarketClient(FakeConnection(append_trailing_byte=True))
    with pytest.raises(RuntimeError, match="exactly 48 bytes"):
        await client.get_accounting_migration_status()
    with pytest.raises(RuntimeError, match="exactly 64 bytes"):
        await client.get_accounting_health()


@pytest.mark.asyncio
async def test_compute_market_rejects_zero_hash_before_submission() -> None:
    connection = FakeConnection()
    client = ComputeMarketClient(connection)
    requester = Keypair.from_seed(bytes(range(32)))
    with pytest.raises(ValueError, match="zero hash"):
        await client.submit_job(requester, 10, 100, bytes(32))
    assert connection.calls == []
